webapp/src/oEmbed.jsx

```diff
@@ -0,0 +1,87 @@
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+
+import {id as pluginId} from './manifest';
+import {getDimensionsForProvider, getOEmbedUrl, getProviderForUrl} from './utils';
+
+export default class OEmbed extends Component {
+    static propTypes = {
+        embed: PropTypes.shape({
+            url: PropTypes.string.isRequired,
+        }),
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.oembedUrl = getOEmbedUrl(this.props.embed.url);
+        const provider = getProviderForUrl(this.props.embed.url);
+        this.dimensions = getDimensionsForProvider(provider.provider_name);
+        this.state = {
+            html: null,
+            thumbnailUrl: null,
+        };
+    }
+
+    componentDidMount() {
+        const options = {
+            method: 'POST',
+            headers: {'Content-Type': 'application/json'},
+            body: this.oembedUrl,
+        };
+
+        fetch(`/plugins/${pluginId}/`, options).
+            then((res) => res.json()).
+            then((res) => {
+                if (res.html) {
+                    this.setState({html: res.html});
+                } else {
+                    this.setState({thumbnailUrl: res.thumbnail_url});
+                }
+            }).
+            catch((err) => {
+                // eslint-disable-next-line no-console
+                console.error(err);
+            });
+    }
+
+    render() {
+        const {html, thumbnailUrl} = this.state;
+
+        if (this.dimensions === null) {
+            return <p>{'Plugin error: Invalid oEmbed provider dimensions'}</p>;
+        }
+
+        const style = {
+            width: `${this.dimensions.width}px`,
+            height: `${this.dimensions.height}px`,
+        };
+
+        const iframeStyle = {
+            ...style,
+            border: 0,
+        };
+
+        let inner;
+        if (html) {
+            if (html.startsWith('<iframe')) {
```

Comment:
Not safe. Consider `<iframe></iframe><script>alert("totally outside the iframe")</script>` or `<iframe src="javascript:alert('still executed in the parent origin')"></iframe>` or `<iframe srcdoc="<script>alert('same-origin with the parent')</script>"></iframe>`. The solution would be to either parse the HTML, validate that there's only one root-level element of type `iframe`, and validate all its attributes; or to simply wrap everything in an iframe regardless of whether there's one already.